File: Antidote/KeyboardNotificationController.m
//
//  KeyboardNotificationController.m
//  Antidote
//
//  Created by Dmytro Vorobiov on 10/09/15.
//  Copyright (c) 2015 dvor. All rights reserved.
//

#import "KeyboardNotificationController.h"

@interface KeyboardNotificationController ()

@end

@implementation KeyboardNotificationController

#pragma mark -  Lifecycle

- (instancetype)init
{
    self = [super init];

    if (! self) {
        return nil;
    }

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillShowNotification:)
                                                 name:UIKeyboardWillShowNotification
                                               object:nil];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillHideNotification:)
                                                 name:UIKeyboardWillHideNotification
                                               object:nil];

    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark -  Public

- (void)keyboardWillShowAnimated:(NSNotification *)keyboardNotification
{
    // nop
}

- (void)keyboardWillHideAnimated:(NSNotification *)keyboardNotification
{
    // nop
}

#pragma mark -  Notifications

- (void)keyboardWillShowNotification:(NSNotification *)notification
{
    weakself;
    [self performAnimatedBlock:^{
        strongself;
        [self keyboardWillShowAnimated:notification];
    } withKeyboardNotification:notification];
}

- (void)keyboardWillHideNotification:(NSNotification *)notification
{
    weakself;
    [self performAnimatedBlock:^{
        strongself;
        [self keyboardWillHideAnimated:notification];
    } withKeyboardNotification:notification];
}

#pragma mark -  Private

- (void)performAnimatedBlock:(void (^)())block withKeyboardNotification:(NSNotification *)notification
{
    NSDictionary *userInfo = [notification userInfo];

    NSTimeInterval duration = [userInfo[UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    UIViewAnimationCurve curve = [userInfo[UIKeyboardAnimationCurveUserInfoKey] integerValue];

    UIViewAnimationOptions options = 0;

    switch (curve) {
        case UIViewAnimationCurveEaseInOut:
            options |= UIViewAnimationOptionCurveEaseInOut;
            break;
        case UIViewAnimationCurveEaseIn:
            options |= UIViewAnimationOptionCurveEaseIn;
            break;
        case UIViewAnimationCurveEaseOut:
            options |= UIViewAnimationOptionCurveEaseOut;
            break;
        case UIViewAnimationCurveLinear:
            options |= UIViewAnimationOptionCurveLinear;
            break;
    }

    [UIView animateWithDuration:duration delay:0.0 options:options animations:block completion:nil];
}

@end

File: Antidote/KeyboardNotificationController.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import UIKit

class KeyboardNotificationController: UIViewController {
    init() {
        super.init(nibName: nil, bundle: nil)

        let center = NotificationCenter.default
        center.addObserver(self, selector: #selector(KeyboardNotificationController.keyboardWillShowNotification(_:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
        center.addObserver(self, selector: #selector(KeyboardNotificationController.keyboardWillHideNotification(_:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
    }

    required convenience init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {
        let center = NotificationCenter.default
        center.removeObserver(self)
    }

    func keyboardWillShowAnimated(keyboardFrame frame: CGRect) {
        // nop
    }

    func keyboardWillHideAnimated(keyboardFrame frame: CGRect) {
        // nop
    }

    @objc func keyboardWillShowNotification(_ notification: Notification) {
        handleNotification(notification, willShow: true)
    }

    @objc func keyboardWillHideNotification(_ notification: Notification) {
        handleNotification(notification, willShow: false)
    }
}

private extension KeyboardNotificationController {
    func handleNotification(_ notification: Notification, willShow: Bool) {
        let userInfo = notification.userInfo!

        let frame = (userInfo[UIKeyboardFrameEndUserInfoKey] as! NSValue).cgRectValue
        let duration = (userInfo[UIKeyboardAnimationDurationUserInfoKey] as! NSNumber).doubleValue
        let curve = UIViewAnimationCurve(rawValue: (userInfo[UIKeyboardAnimationCurveUserInfoKey] as! NSNumber).intValue)!

        let options: UIViewAnimationOptions

        switch curve {
            case .easeInOut:
                options = UIViewAnimationOptions()
            case .easeIn:
                options = .curveEaseIn
            case .easeOut:
                options = .curveEaseOut
            case .linear:
                options = .curveLinear
        }

        UIView.animate(withDuration: duration, delay: 0.0, options: options, animations: { [unowned self] in
            willShow ? self.keyboardWillShowAnimated(keyboardFrame: frame) : self.keyboardWillHideAnimated(keyboardFrame: frame)
        }, completion: nil)
    }
}
